src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Unix.cs

```diff
@@ -433,69 +434,64 @@ public static void MoveDirectory(string sourceFullPath, string destFullPath)
 
         public static void RemoveDirectory(string fullPath, bool recursive)
         {
-            var di = new DirectoryInfo(fullPath);
-            if (!di.Exists)
+            if (!TryRemoveDirectory(fullPath, topLevel: true, throwWhenNotEmpty: !recursive))
             {
-                throw Interop.GetExceptionForIoErrno(Interop.Error.ENOENT.Info(), fullPath, isDirectory: true);
+                RemoveDirectoryRecursive(fullPath);
             }
-            RemoveDirectoryInternal(di, recursive, throwOnTopLevelDirectoryNotFound: true);
         }
 
-        private static void RemoveDirectoryInternal(DirectoryInfo directory, bool recursive, bool throwOnTopLevelDirectoryNotFound)
+        private static void RemoveDirectoryRecursive(string fullPath)
         {
             Exception? firstException = null;
 
-            if ((directory.Attributes & FileAttributes.ReparsePoint) != 0)
+            try
             {
-                DeleteFile(directory.FullName);
-                return;
-            }
+                foreach ((string childPath, bool isDirectory) in
+                            new FileSystemEnumerable<(string, bool)>(
+                                fullPath,
+                                static (ref FileSystemEntry entry) =>
+                                {
+                                    // Don't report symlinks to directories as directories.
+                                    bool isRealDirectory = !entry.IsSymbolicLink && entry.IsDirectory;
 
-            if (recursive)
-            {
-                try
+                                    return (entry.ToFullPath(), isRealDirectory);
+                                },
+                                EnumerationOptions.Compatible))
```

Comment:
Tiny nit: Could you please save the `FileSystemEnumerable` instance into a variable, outside of the try? It won't enumerate until you enter the `foreach`, and it's easier to read and debug:

Note: GitHub is not allowing me to add this as a suggestion 😠 (maybe it's a very complicated diff?).
```cs
            FileSystemEnumerable<(string, bool)> entries = new(
                fullPath,
                static (ref FileSystemEntry entry) =>
                {
                        // Don't report symlinks to directories as directories.
                        bool isRealDirectory = !entry.IsSymbolicLink && entry.IsDirectory;
                    return (entry.ToFullPath(), isRealDirectory);
                },
                EnumerationOptions.Compatible);

            try
            {
                foreach ((string childPath, bool isDirectory) in entries)
```